Add sops lookup plugin

endorama opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Add a new lookup plugin for using [`sops`](https://github.com/mozilla/sops) to lookup content from sops-encrypted files.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes #47770

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
plugin 

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
Sops can be used in place of Ansible Vault as an easy to manage at rest encryption for secrets in the repository.

Sops is able to encrypt entire files, this plugin allow making advantage of that functionality and load as strings sops encrypted files (think of TLS cert keys, SSH keys or any other entire file secret you may need).

This plugin require the `sops` command to be available in the path.

Related to #36982

endorama committed:
add sops lookup plugin

jwitko commented on lib/ansible/plugins/lookup/sops.py:
```diff
@@ -0,0 +1,154 @@
+            try:
+                if lookupfile:
+                    # Run sops directly, python module is deprecated
+                    command = ["sops", "--decrypt", encrypted_file]
```
This variable is causing an `undefined variable` error.
```suggestion
                    command = ["sops", "--decrypt", lookupfile]
```

jwitko replied:
@endorama ping

endorama replied:
I pushed the fix!

jwitko reviewed (changes requested):
I went looking hoping that I'd find something just like this.  Thank you very much for contributing!  I have tested this and was able to make it work quite nice with the suggested edit below, although I did have to `| from_json` in order to get usable output.  Not sure if its desirable to handle that in the module or let the user manipulate the output?

Example:
Play:
```
---
- hosts: local
  gather_facts: False
  tasks:
    - copy:
        content: "{{ lookup('sops', 'enc.json') }}"
        dest: /tmp/tmp
```

output:
```
$ cat /tmp/tmp
b'{\n\t"auths": {\n\t\t"iad.ocir.io": {\n\t\t\t"username": "TEST/TEST",\n\t\t\t"password": "nn\\u003c+0xX55V;V\\u003eRjusAew",\n\t\t\t"email": "redacted@email.com",\n\t\t\t"auth": "TEST"\n\t\t}\n\t}\n}'
```

output when adding `| from_json` to the `content:` line:
```
{"auths": {"iad.ocir.io": {"username": "TEST/TEST", "password": "nn<+0xX55V;V>RjusAew", "email": "redacted@email.com", "auth": "TEST"}}}
```

You can see how the output is completely unusable without `| from_json` so I'm kind of having a hard time thinking of a reason why we wouldn't want it parsed to a usable format from inside the module?

endorama committed:
fix pylint

endorama committed:
fix undefined encrypted_file variable

endorama commented:
@jwitko I fixed pylint and the undefined var. (I'll squash commits after review if you are ok)

Regaring the json not being parsed, this is probably a bug.

Let me clarify this a bit. Sops can operate in 2 "modes": structured or unstructured file.  
Structured files are json and yaml, you expect that to be parsed before use.  
Unstructured files are any binary or textual file that you expect to use as is after decryption.

I imagined the lookup plugin to be useful for unstructured files mainly. For example certificates private keys. In this case you use the output as it is, and you expect no transformation to happen.

Instead, you could use the vars plugin, proposed in PR #59641, to load playbook variables from structured files variables.

Your particular use case seems to be broken by some formatting issue in the string read from the sops output. I replicated the behaviour and trying to fix it, as it should not happen.

endorama committed:
decode sops output as text by default

endorama committed:
add variable to control decrypted content print in logs

endorama committed:
use Sops class decryption method

endorama commented:
Commit 0a5924d fixes the issue showed by @jwitko, which I guess was due to changes in how `process.communicate()` handles command output (maybe Python 3 related?)

I also added 2 plugin variables:
- `binary` to control if the `sops` command output should handled as binary or text (default text)
- `print` to control if the decrypted secret should be printed on screen or not (instead of commenting/decommenting the code)

sivel commented on lib/ansible/plugins/lookup/sops.py:
```diff
@@ -0,0 +1,171 @@
+                if lookupfile:
+                    (output, err, exit_code) = Sops.decrypt(lookupfile)
+
+                    if sops_text_output:
```
Lookups should always return "text".  As such, you should make use of `ansible.module_utils._text.to_text` to handle enforce the type.
```
output = to_text(output, errors='surrogate_or_strict')
```

endorama replied:
thanks for the suggestion! I changed the code to use the proper ansible function.

endorama committed:
lookup should return text, use appropriate ansible facility

endorama commented:
ready_for_review

jwitko commented:
@sivel Not sure how to make the CI retry but it looks like it failed due to a transient issue in the tests not an actual issue in the code.

endorama commented:
Shall I squash this in a single commit?

cameron-elliott commented:
Hey! I just checked out this PR, which means it must be pretty cool, because I struggled a fair bit being new to gpg, sops, ansible and Python. :smile:
If I can provide any feedback or newbie-level help, let me know, but I think having this available is pretty neato, thanks for contributing it!

jwitko commented:
@sivel seems we're a but stuck here

mhumeSF commented:
Am I running this correctly? Using the code from the example in `sops.py`

```
---
  - name: output secrets to screen (BAD IDEA)
    debug:
        msg: "Content: {{lookup('sops', item)}}"
    with_file:
    - secrets.enc.yaml
```
Getting errors when running:
```
Sops lookup using None as file
fatal: [localhost]: FAILED! => {
    "msg": "An unhandled exception occurred while running the lookup plugin 'sops'. Error was a <class 'ansible.errors.AnsibleError'>, original message: could not locate file in lookup: password: ENC[AES256_GCM,data:redacted,tag:redacted,type:str] sops: kms: -   arn: arn:aws:kms:<redacted> created_at: '2019-10-10T21:18:01Z' enc: redacted aws_profile: "" gcp_kms: [] azure_kv: [] lastmodified: '2019-10-10T21:18:02Z' mac: ENC[AES256_GCM,data:redacted,iv:redacted,tag:redacted,type:str] pgp: [] unencrypted_suffix: _unencrypted version: 3.4.0
```

endorama committed:
use ansible.module_utils._text.to_native

As required by Ansible documentation on [raising errors][raising-errors]
from plugins, use to_native to wrap errors to ensure string compatibility
between Python versions.

[raising-errors]: https://docs.ansible.com/ansible/latest/dev_guide/developing_plugins.html#id3

endorama committed:
use with_items instead of with_file in sops lookup documentation

[with_file][with-file], per Ansible documentation, returns the content of
the file. As sops is not able to decrypt a string by itself but requires
the file is passed as argument, passing the content breaks the lookup
plugin as reported by [here][bug-report].

[with_items][with-items] should be used instead.

[with-file]: https://docs.ansible.com/ansible/2.4/playbooks_loops.html#looping-over-files
[with-items]: https://docs.ansible.com/ansible/2.4/playbooks_loops.html#standard-loops
[bug-report]: https://github.com/ansible/ansible/pull/59639#issuecomment-540803722

endorama commented:
@mhumeSF thanks for spotting this! That's an error in the documentation.

It uses `with_file`, but as it returns the file *content* `sops` breaks, thus the error you see. I fixed the documentation, but using `with_items` (so passing the *path* to the lookup) should make it work.

mhumeSF commented on lib/ansible/plugins/lookup/sops.py:
```diff
@@ -0,0 +1,162 @@
+    def decrypt(encrypted_file):
+        # Run sops directly, python module is deprecated
+        command = ["sops", "--decrypt", encrypted_file]
+        process = Popen(command, stdout=PIPE, stderr=PIPE)
```
Would this be better? `process = Popen(command, stdout=PIPE, stderr=PIPE, encoding='utf8')`

mhumeSF replied:
This was due to an environment issue on my end, so disregard.

mhumeSF commented:
@endorama Updated with `with_items`, which helped. Now the file was found but, was getting a new error of `Error was a <class 'TypeError'>, original message: a bytes-like object is required, not 'str'"`.

mhumeSF commented:
@sivel These two plugins are working great. What else would be needed to get this and https://github.com/ansible/ansible/pull/59641 merged?

mhumeSF reviewed (approved):
👍

endorama committed:
uniform sops exception handling between plugins

endorama commented:
I'm failing to see where it breaks, any hint appreciated! (Maybe rebase onto devel would help?)

felixfontein commented on lib/ansible/plugins/lookup/sops.py:
```diff
@@ -0,0 +1,171 @@
+DOCUMENTATION = """
+    lookup: sops
+    author: Edoardo Tenani <e.tenani@arduino.cc>
+    version_added: "2.8"
```
```suggestion
    version_added: "2.10"
```

felixfontein commented on lib/ansible/plugins/lookup/sops.py:
```diff
@@ -0,0 +1,171 @@
+    short_description: Read sops encrypted file contents
+    description:
+        - This lookup returns the contents from a file on the Ansible controller's file system.
+        - This lookup requires the sops executable to be available in the controller PATH.
```
```suggestion
        - This lookup requires the C(sops) executable to be available in the controller PATH.
```

felixfontein commented on lib/ansible/plugins/lookup/sops.py:
```diff
@@ -0,0 +1,171 @@
+            required: False
+            default: False
+    notes:
+        - this lookup does not understand 'globing' - use the fileglob lookup instead.
```
```suggestion
        - This lookup does not understand 'globing' - use the fileglob lookup instead.
```

mweinelt replied:
Globing -> Globbing?

felixfontein replied:
Indeed :) I totally missed that...

endorama replied:
```suggestion
        - This lookup does not understand 'globbing' - use the fileglob lookup instead.
```

felixfontein commented on lib/ansible/plugins/lookup/sops.py:
```diff
@@ -0,0 +1,171 @@
+
+# From https://github.com/mozilla/sops/blob/master/cmd/sops/codes/codes.go
+# Should be manually updated
+sops_error_codes = {
```
You want to move that to module_utils once at least one of this and #59641 is merged.

endorama replied:
I will surely need some guidance on that, as I already tried but couldn't make it work. But sure, happy to tackle that.

felixfontein replied:
Essentially, it works as follows:
 1. Create a file `lib/ansible/module_utils/sops.py` which starts with the following, and add the code which is used by both plugins (i.e. `sops_error_codes`, `SopsError`, `Sops` from this PR):
```
# Copyright (c), xxx <yyy@zzz.com>, 2018-2020
# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)

from __future__ import absolute_import, division, print_function
```
2. Add `from ansible.module_utils.sops import Sops, SopsError` to the plugin (and remove the code you've copied).

endorama replied:
Would you prefer this change to be in another PR or it's fine to add it to this one?

felixfontein replied:
I think it will be faster to merge this PR (since the other has the paths issue), so let's make the change here and get this merged, and adjust the other PR according (once this is merged). Does that sound good for you?

felixfontein commented on lib/ansible/plugins/lookup/sops.py:
```diff
@@ -0,0 +1,171 @@
+        _terms:
+            description: path(s) of files to read
+            required: True
+        print:
```
What is this option needed for? After all, these are secrets you're dealing with, and if the user wants to print them she can use the `debug` module.

endorama replied:
You're perfectly right, actually I didn't thought about that. I'll remove it.

endorama replied:
As suggested, see a262523

endorama committed:
Apply suggestions from code review

Co-Authored-By: Felix Fontein <felix@fontein.de>

endorama committed:
remove sops lookup plugin print option

Is no longer possible to print the decrypted secrets directly from this
plugin, but `debug` module can be used instead.

felixfontein commented on lib/ansible/plugins/lookup/sops.py:
```diff
@@ -0,0 +1,157 @@
+
+DOCUMENTATION = """
+    lookup: sops
+    author: Edoardo Tenani <e.tenani@arduino.cc>
```
```suggestion
    author: Edoardo Tenani (@endorama) <e.tenani@arduino.cc>
```
(Same for the other PR. This will make sure that you will be notified about issues / PRs for this plugin by ansibot.)

felixfontein reviewed (commented):
Besides this and creating a module_utils, you should also add tests for this. As an example, you can look at:
- https://github.com/ansible/ansible/tree/devel/test/integration/targets/lookup_rabbitmq
- https://github.com/ansible/ansible/tree/devel/test/integration/targets/setup_rabbitmq

(You will be able to re-use `setup_sops` for #59641.)

endorama committed:
add github handle to author

endorama committed:
add lookup_sops integration tests

endorama committed:
extract sops module

endorama committed:
use sops module

endorama committed:
add setup_sops integration target

endorama commented:
@felixfontein I added the tests, I'm uncertain on which cases should I test (as testing `sops` behaviour itself seems redundant); I added a "missing file", "wrong file" and "success" tests, let me know if you were thinking of some other test cases.

felixfontein commented on lib/ansible/module_utils/sops.py:
```diff
@@ -0,0 +1,60 @@
+
+from __future__ import absolute_import, division, print_function
+
+from ansible.errors import AnsibleError
```
Hmm, this is in fact a problem. Everything in `module_utils` can only use stuff from `module_utils`, but not from the rest of Ansible. That's somewhat unfortunate.

How about having a function `createSopsErrorMessage()` instead which is used by `SopsError` in the plugin to create the message? Then the heavy part (`sops_error_codes` and the logic to use it) is in `module_utils`, but `module_utils/sops.py` doesn't need to know about `AnsibleError`.

felixfontein commented:
Please remove the binary file from the PR.

Also, you need a file `test/integration/targets/lookup_sops/aliases`, otherwise your test won't be run in CI. (See also the related `test/integration/targets/lookup_sops/aliases:0:0: missing alias 'shippable/posix/group[1-5]' or 'unsupported'` error above.)

felixfontein commented:
I think the PR in ansible-collections/community.general#374 is now finished. Anyone interested in it, please take a look and check it out! Reviews and testing are always very welcome :)

felixfontein commented:
The sops lookup plugin will be part of the next community.general release.

close_me

endorama commented:
In case someone subscribed to this PR is interested, I started the process to create a new collection named `community.sops` where to move all sops related plugins and code. In case someone is interested in being a maintainer (I will) please get in touch in https://github.com/ansible-collections/overview/issues/85
